vendor/github.com/danwinship/knftables/types.go

```diff
@@ -0,0 +1,338 @@
+	TypeOf string
+
+	// Flags are the set flags
+	Flags []SetFlag
+
+	// Timeout is the time that an element will stay in the set before being removed.
+	// (Optional; mandatory for sets that will be added to from the packet path)
+	Timeout *time.Duration
+
+	// GCInterval is the interval at which timed-out elements will be removed from the
+	// set. (Optional; FIXME DEFAULT)
+	GCInterval *time.Duration
+
+	// Size if the maximum numer of elements in the set.
+	// (Optional; mandatory for sets that will be added to from the packet path)
+	Size *uint64
+
+	// Policy is the FIXME
+	Policy *SetPolicy
+
+	// AutoMerge indicates that adjacent/overlapping set elements should be merged
+	// together (only for interval sets)
+	AutoMerge *bool
+
+	// Comment is an optional comment for the object.
+	Comment *string
+
+	// Handle is an identifier that can be used to uniquely identify an object when
+	// deleting it. When adding a new object, this must be nil
+	Handle *int
+}
+
+// Map represents the definition of an nftables map (but not its elements)
+type Map struct {
+	// Name is the name of the map.
+	Name string
+
+	// Type is the type of the map key and value (eg "ipv4_addr : verdict"). Either
+	// Type or TypeOf, but not both, must be non-empty.
+	Type string
+
+	// TypeOf is the type of the set key as an nftables expression (eg "ip saddr : verdict").
+	// Either Type or TypeOf, but not both, must be non-empty.
+	TypeOf string
+
+	// Flags are the map flags
+	Flags []SetFlag
+
+	// Timeout is the time that an element will stay in the set before being removed.
+	// (Optional; mandatory for sets that will be added to from the packet path)
+	Timeout *time.Duration
+
+	// GCInterval is the interval at which timed-out elements will be removed from the
+	// set. (Optional; FIXME DEFAULT)
+	GCInterval *time.Duration
+
+	// Size if the maximum numer of elements in the set.
+	// (Optional; mandatory for sets that will be added to from the packet path)
+	Size *uint64
+
+	// Policy is the FIXME
+	Policy *SetPolicy
+
+	// Comment is an optional comment for the object.
+	Comment *string
+
+	// Handle is an identifier that can be used to uniquely identify an object when
+	// deleting it. When adding a new object, this must be nil
+	Handle *int
+}
+
+// Element represents a set or map element
+type Element struct {
+	// Set is the name of the set that contains this element (or the empty string if
+	// this is a map element.)
+	Set string
+
+	// Map is the name of the map that contains this element (or the empty string if
+	// this is a set element.)
+	Map string
+
+	// Key is the element key. (The list contains a single element for "simple" keys,
+	// or multiple elements for concatenations.)
+	Key []string
+
+	// Value is the map element value. As with Key, this may be a single value or
+	// multiple. For set elements, this must be nil.
+	Value []string
+
+	// Comment is an optional comment for the element
+	Comment *string
+}
```

Comment:
@danwinship 
should we also add a `Elements` type ?
```go
type Elements []Element
```

this would allow to batch all IPs/ports together in one NFT command, reducing total-number of bytes to write and may even improve perf .. as less rules to be parse in total 
```bash
# Element.writeOperation
add element ip kube-proxy firewall { 10.1.2.3 . tcp . 80 }
add element ip kube-proxy firewall { 10.1.2.4 . tcp . 80 }
add element ip kube-proxy firewall { 10.1.2.5 . tcp . 80 }
```

```bash
# Elements.writeOperation
add element ip kube-proxy firewall { 10.1.2.3 . tcp . 80,  10.1.2.4 . tcp . 80, 10.1.2.5 . tcp . 80 }
```

Strictly following the userspace nft API the add element command should accept multiple keys but that won't look neat. 
```go
tx.Add(&knftables.Element{
        Set: "firewall",
        Key: [][]string{{"10.1.2.3", "tcp", "80"},{"10.1.2.4", "tcp", "80"},{"10.1.2.5", "tcp", "80"}},
})
```

Do you think we should try either of the approach to batch the elements of set/map together ?

Reply:
if you want to discuss the knftables API, it's probably better to file issues there?

> this would allow to batch all IPs/ports together in one NFT command, reducing total-number of bytes to write and may even improve perf .. as less rules to be parse in total

eventually we'll be doing partial syncs like in iptables so there'll be a lot less parsing anyway.

Anyway, for most sets, we generally don't know more than 1 (or "a few") elements at a time, because we figure everything out on a service-by-service basis. So we'd have to do a lot of reorganizing to be able to take advantage of something like this.